api/client/samples/analogous_years/lib/final_ranks_computation.py

```diff
@@ -0,0 +1,259 @@
+    :param final_date: A date in YYYY-MM-DD format
+    :param methods_list: a sublist of ['cumulative', 'euclidean', 'ts-features', 'dtw']
+    :param all_ranks: Boolean to determine if all ranks will be displayed or a composite rank
+    :param weights: Float determining the weight given to each entity
+    :param enso: Boolean to include ENSO
+    :param enso_weight: Float
+    :param provided_start_date: A string in YYYY-MM-DD format
+    :return: A tuple (string, dataframe)
+    The string contains '_' separated region, item, date
+    The dataframe contains integer values (ranks)
+    """
+    logger = client.get_logger()
+    combined_items_distances = None
+    entities = common_start_date(client, entities, provided_start_date)['entities']
+    start_date = common_start_date(client, entities, provided_start_date)['start_date']
+    if not weights:
+        weights = [1] * len(entities)
+    if enso:
+        entities.append(enso_data(start_date))
+        if enso_weight:
+            weights.append(enso_weight)
+        else:
+            weights.append(1)
+    file_name = get_file_name(client, entities, initial_date, final_date)
+    for i in range(len(entities)):
+        gro_item = client.lookup('items', entities[i]['item_id'])['name']
+        combined_methods_distances_df = combined_methods_distances(
+            ranked_df_dictionary(
+                client, entities[i], initial_date,
+                final_date, gro_item, methods_list))
+        numpy_combined_methods_distances = combined_methods_distances_df.values
+        if combined_items_distances is None:
+            combined_items_distances = np.zeros(numpy_combined_methods_distances.shape)
+        combined_items_distances = combined_items_distances + np.power(
+            weights[i] * numpy_combined_methods_distances, 2)
+    combined_items_distances = pd.DataFrame(np.sqrt(combined_items_distances),
+                                            index=combined_methods_distances_df.index,
+                                            columns=combined_methods_distances_df.columns)
+    combined_items_distances['composite'] = combined_items_distances.sum(axis=1, skipna=True)
+    ranks = []
+    for column_name in combined_items_distances.columns:
+        combined_items_distances.sort_values(by=column_name, inplace=True)
+        column_new_name = column_name.split('_')[0] + '_rank'
+        combined_items_distances[column_new_name] = \
+            combined_items_distances.reset_index().index + 1
+        ranks.append(column_new_name)
+        combined_items_distances.sort_index(inplace=True)
+    if all_ranks:
+        display_dataframe = combined_items_distances[ranks]
+    else:
+        display_dataframe = combined_items_distances['composite_rank']
+    logger.info("\n Computing analogous years ranks for {} \n".format(file_name))
+    return file_name, display_dataframe
+
+
+def save_to_csv(dataframe_file_name, logger, all_ranks=None, report=True, output_dir=''):
```

Comment:
csv and png output should be two different functions

Reply:
Done